pkg/controller/directvolumemigration/rsync.go

```diff
@@ -1651,3 +1678,625 @@ func (t *Task) deleteProgressReportingCRs(client k8sclient.Client) error {
+	// password password used by startup Rsync command
+	password string
+	// privileged whether Rsync Pod will run privileged
+	privileged bool
+	// resourceReq resource requirements for the Pod
+	resourceReq corev1.ResourceRequirements
+	// nodeName node on which Rsync Pod will be launched
+	nodeName string
+	// destIP destination IP address for Stunnel route
+	destIP string
+	// rsyncOptions rsync command to execute
+	rsyncOptions []string
+}
+
+// getRsyncClientPodTemplate given RsyncClientPodRequirements, returns a Pod template
+func (req *rsyncClientPodRequirements) getRsyncClientPodTemplate() corev1.Pod {
+	runAsUser := int64(0)
+	trueBool := true
+	volumes := []corev1.Volume{}
+	volumeMounts := []corev1.VolumeMount{}
+	containers := []corev1.Container{}
+	volumeMounts = append(volumeMounts, corev1.VolumeMount{
+		Name:      req.pvInfo.name,
+		MountPath: fmt.Sprintf("/mnt/%s/%s", req.namespace, req.pvInfo.name),
+	})
+	volumes = append(volumes, corev1.Volume{
+		Name: req.pvInfo.name,
+		VolumeSource: corev1.VolumeSource{
+			PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+				ClaimName: req.pvInfo.name,
+			},
+		},
+	})
+	rsyncCommand := []string{"rsync"}
+	rsyncCommand = append(rsyncCommand, req.rsyncOptions...)
+	rsyncCommand = append(rsyncCommand, fmt.Sprintf("/mnt/%s/%s/", req.namespace, req.pvInfo.name))
+	rsyncCommand = append(rsyncCommand, fmt.Sprintf("rsync://root@%s/%s", req.destIP, req.pvInfo.name))
+	labels := map[string]string{
+		"app":                   DirectVolumeMigrationRsyncTransfer,
+		"directvolumemigration": DirectVolumeMigrationRsyncClient,
+	}
+	labels = Union(labels, GetRsyncPodSelector(req.pvInfo.name))
+	containers = append(containers, corev1.Container{
+		Name:  DirectVolumeMigrationRsyncClient,
+		Image: req.image,
+		Env: []corev1.EnvVar{
+			{
+				Name:  "RSYNC_PASSWORD",
+				Value: req.password,
+			},
+		},
+		TerminationMessagePolicy: corev1.TerminationMessageFallbackToLogsOnError,
+		Command:                  rsyncCommand,
+		Ports: []corev1.ContainerPort{
+			{
+				Name:          DirectVolumeMigrationRsyncClient,
+				Protocol:      corev1.ProtocolTCP,
+				ContainerPort: int32(22),
+			},
+		},
+		VolumeMounts: volumeMounts,
+		SecurityContext: &corev1.SecurityContext{
+			Privileged:             &req.privileged,
+			RunAsUser:              &runAsUser,
+			ReadOnlyRootFilesystem: &trueBool,
+		},
+		Resources: req.resourceReq,
+	})
+	clientPod := corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "dvm-rsync-",
+			Namespace:    req.namespace,
+			Labels:       labels,
+		},
+		Spec: corev1.PodSpec{
+			RestartPolicy: corev1.RestartPolicyNever,
+			Volumes:       volumes,
+			Containers:    containers,
+			NodeName:      req.nodeName,
+			SecurityContext: &corev1.PodSecurityContext{
+				SupplementalGroups: req.pvInfo.supplementalGroups,
+				FSGroup:            req.pvInfo.fsGroup,
+				SELinuxOptions:     req.pvInfo.seLinuxOptions,
+			},
+		},
+	}
+	return clientPod
```

Comment:
this looks like a perfect use case for templates, wonder if we should define a YAML file/variable, read and parse the variables through the template?

Reply:
@alaypatel07 I investigated whether we can use template here...we can. I feel like the gain of readability doesnt justify in this case because we need templates for containers, volumes, and the pods at the very least since they are all dynamically created. Maintaining three templates, executing them and unmarshling them back into the Pod object seems unnecessary.

Reply:
@pranavgaikwad I would expect only 1 pod template and all the variable parts will be filled as variables from the struct. That gets rid of all the verbosity of the existing code.

Reply:
If you still feel it's not something we want I'm fine.